### PR TITLE
Add support for Protocol Info objects

### LINKED ISSUE #41: AMQP Message Exchange
I noticed there is no construct for an exchange in AMQP. It would be nice to have that explicitly called out somehow.

### LINKED ISSUE #42: Consideration: Protocol specific message objects
AMQP, STOMP, MQTT, and others all have variations in the message structure.

I wanted to suggest that you consider the idea of protocol specific message objects *or* protocol specific message object properties, that embrace the individuality and uniqueness of each protocol's message specification.

In addition, the variations between AMQP 0.9 and 1.0 are enough to warrant protocol version disambiguation.

Have you previously put any thought into this? As someone who's a fairly invested in AMQP 0.9, I'd love to see more detail in that area than what the spec currently provides.

### LINKED ISSUE #164: [FEATURE REQUEST] Message-classes/Topic-classes
**Is your feature request related to a problem? Please describe.**
I believe it would be useful to allow protocol specific settings for topics or messages, much like the Kubernetes spec does for persistent volumes in the form of Storage-classes.

In this case I would like to request a feature for publishing messages directly to a queue for the AMQP protocol. As well as the possibility to add a default queue and exchange. Allowing this would cover many desirable use-cases.

**Can't it be tackled using specification extensions?**
I believe that if you support a protocol you should support all its features. So in this case it would not suffice.

**Describe the solution you'd like**
I would suggest something similar to storage-classes in Kubernetes. Think of it like a message-class/topic-class. Adding a layer of abstraction so the topic-class can be reused.

Below I've added an example for topic-classes something similar could be added for messages if desirable. In it I've added a `topic-classes` attribute to the components property. It contains a `default-class` which defines a `protocol` which to me is akin to a `provisioner` in Kubernetes storage-classes. The class states it is the default-class for all topics however, topics can override it by specifying their preferred class.
The `default-class` specifies a `queue` which applies to both subscribing and publishing, unless the queue is overridden in the `subscribe` or `publish` properties of the class. If the properties do not specify a queue the default is used.

```
asyncapi: '1.3.0'
info:
  title: Example API
  version: '1.0.0'
  description: |
    Example API.
  license:
    name: Apache 2.0
    url: 'https://www.apache.org/licenses/LICENSE-2.0'

servers:
- url: some.url.net
  scheme: amqp
  description: Example broker
  variables:
    port:
      description: Secure connection (TLS) is available through port 8883.
      default: '1883'
      enum:
      - '1883'
      - '8883'

topics:
  example-topic:
    class: 
      $ref: '#/components/topic-classes/default-class'
    publish:
      $ref: '#/components/messages/example'
  example-message-class-topic:
    publish:
      $ref: '#/components/messages/message-class-example'

components:
  messages:
    example:
      summary: An example message.
      payload:
        $ref: "#/components/schemas/example"
    message-class-example:
      summary: An example message.
      class: '#/components/message-classes/rpc-class'
        $ref: 
      payload:
        $ref: "#/components/schemas/example"
  schemas:
    example:
      type: object
      properties:
        example:
          type: sting
          description: An example string
  topic-classes:
    default-class: 
      scheme: amqp
      default: true
      queue:
        name: example-queue
      publish: 
        direct-to-queue: true
        routing-key: routed-to-something
#      exchange: <- Defining exchange specifics would also be possible, just not when publishing directly to a queue.
#        name: overriding-exchange 
#        type: fanout, direct, topic
      subscribe:
        binding-key: routed-to-something-else
        queue:
          name: overriding-queue
  message-classes:
    rpc-class:
      scheme: ampq
      delivery_mode: 2
      content_type: application/json
      reply_to: rpc-queue
```

**Describe alternatives you've considered**
I've considered adding extensions which tools could read. Though I feel it would not be a proper solution due to the reasons I've stated above. If a specification supports a protocol it should support the protocols features.

**Additional context**
Some benefits of topic-classes:

1. Adding topic-classes/message-classes opens up the specification to many more protocol-specific use-cases.
2. Topic-classes would not be a breaking change, it would simply add functionality as new properties. 
3. I believe this proposal would tie into https://github.com/asyncapi/asyncapi/issues/42 as it has a few commonalities.
4. Topic-classes/message-classes can be reused across multiple topics/messages.
5. Message-classes could be used to implement features like [RPC in AMQP](https://www.rabbitmq.com/tutorials/tutorial-six-javascript.html). I believe this is similar to the feature-request in https://github.com/asyncapi/asyncapi/issues/94 not certain though.

Let me know what you think.

### PR DESCRIPTION
**Pull request by fmvilas**
### What?
Fixes #41
Fixes #42
Fixes #164

### Proposal
ProtocolInfo objects are opaque objects that may be used by tools to perform/add protocol-specific operations/features. Although the object is opaque from a specification point of view, we'll maintain a catalog of protocolInfo objects, so they are standardized across tools but don't compromise the specification. That doesn't stop anyone to use custom protocolInfo objects, at the expense of creating the tooling to support them.

ProtocolInfo objects can be attached to:
* Channels: Describes protocol-specific information about the channel, e.g., queue name, exchange name or partition number.
* Operations: Describes protocol-specific information about the operation. Example: `noAck` (AMQP).
* Messages: Describes protocol-specific information about the message. Examples: protocol headers/properties like `QoS` (MQTT), `delivery_mode` (AMQP), message `key` (Kafka), etc.

Reusing protocolInfo objects will be possible through traits (#108) and overlays (#107).

### Missing things

- [x] ~Add to Markdown version of the specification.~
- [x] ~Add to JSON Schema version of the specification.~

### CODE REVIEW

**Comment by fmvilas** [comment]
@RobertDiebels @gmr ☝️ early draft of the new feature to support protocol-specific information.

File: examples/next/rpc-client.yml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+      - name: queue
+        schema:
+          type: string
+          pattern: '^amq\\.gen\\-.+$'
+    protocolObjects:
+      amqp-0-9-1:
+        channelIsQueue: true
+        queue:
+          randomName: true
+          exclusive: true
```
Comment (fmvilas) [inline examples/next/rpc-client.yml]: Please, ignore the "quality" of the `amqp-0-9-1` object here. The definition of a specific protocol object must be done separately. This example is just to illustrate the feature.

File: examples/next/rpc-client.yml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    subscribe:
+      operationId: receiveSumResult
+      correlationId:
+        location: $message.header#/correlation_id
+      protocolObjects:
+        amqp-0-9-1:
+          noAck: true
+          properties:
+            correlation_id:
+              type: string
```
Comment (fmvilas) [inline examples/next/rpc-client.yml]: Please, ignore the "quality" of the `amqp-0-9-1` object here. The definition of a specific protocol object must be done separately. This example is just to illustrate the feature.

**Comment by RobertDiebels** [comment]
> 
> 
> @RobertDiebels @gmr ☝️ early draft of the new feature to support protocol-specific information.

+1 for the traits compatibility. Re-using protocol-info seems super useful.

**Comment by fmvilas** [comment]
BTW, there's already a proposal for traits: https://github.com/asyncapi/asyncapi/issues/108.

**Pull request merged**